src/Server/Stream.js

```diff
@@ -0,0 +1,83 @@
+/**
+ * @constructor
+ * @param {Element}
+ */
+function Stream (element) {
+	this.stack = [element]
+	Readable.call(this)
+}
+/**
+ * @type {Object}
+ */
+Stream.prototype = Object.create(Readable.prototype, {
+	/**
+	 * @return {void}
+	 */
+	_read: {value: function read () {
+		this.push(this.stack.length ? toChunk(this.stack.pop(), this.stack) : null)
+	}}
+})
+
+/**
+ * @param {function=}
+ * @return {Stream}
+ */
+function toStream (callback) {
+	var readable = new Stream(this)
+
+	if (typeof callback === 'function')
+		readable.on('end', callback)
+
+	return readable
+}
+
+/**
+ * @param {Element} element
+ * @param {Array} stack
+ * @return {string}
+ */
+function toChunk (element, stack) {
+	while (element.flag === ElementComponent)
+		element = (componentMount(element), element.children)		
+
+	var type = element.type
+	var children = element.children
+	var length = children.length
+	var output = ''
+
+	switch (element.flag) {
+		case ElementPromise:
+			return void element.type.then(function (element) {
```

Comment:
this seems incorrect. `toStream` is basically synchronous and even if it was async the promise isn't used to control flow

Reply:
toStream is asynchronous, the `_read` method of `Stream` is called asynchronously by the Readable class that Stream extends, this `toChunk` function returns a small chunk of the output when that happens until there is nothing left to return. 

This branch is only used when the element is a Promise type, i.e. 

```js
async render(){
    return await ...
}
```

In such a case we delegate a calling toChunk again when the promise resolve with the resolved value of the promise. At the end of this function when we call `writable.push(output)` is when we yield back control to the Readable Stream that handles async control flow.

Reply:
but the promise resolves out of band
so either the stream emits

```
<html></html>
<div>promised</div>
```

or the promise is never guaranteed to resolve (this is what I think is happening) and the `toChunk()` is called on an ended stream

Reply:
Since we control when to yield control back to the Readable, the next call to `_read` is never called until we call `writable.push` so it is never out of sync. I think the [node Readable streams](https://nodejs.org/api/stream.html#stream_readable_streams) documentation has more information on how this works.